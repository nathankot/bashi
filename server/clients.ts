import * as r from "redis";
import { Configuration, OpenAIApi } from "openai";

import "https://deno.land/x/dotenv/load.ts";

const whisperEndpoint = Deno.env.get("WHISPER_TRANSCRIBE_ENDPOINT");
if (whisperEndpoint == null || whisperEndpoint.length === 0) {
  throw new Error("WHISPER_TRANSCRIBE_ENDPOINT must be declared");
}

export const redis = r.createClient({
  url: Deno.env.get("REDIS_URL"),
});

export const openai = new OpenAIApi(
  new Configuration({
    apiKey: Deno.env.get("OPENAI_KEY"),
  })
);

export const whisper = {
  async transcribe(audio: ArrayBuffer) {
    const whisperRequest = new FormData();
    whisperRequest.append("audio", new Blob([audio]));
    const whisperResponse = await fetch(whisperEndpoint + "?language=en", {
      method: "POST",
      body: whisperRequest,
    });
    const whisperBody = await whisperResponse.json();
    const text = whisperBody.text;
    if (typeof text !== "string") {
      throw new Error("expected whisper response body to be string");
    }
    return text;
  },
};

export async function setup() {
  console.log("connecting redis client");
  await redis.connect();
}

export async function gracefulStop() {
  console.log("disconnecting redis");
  const rc = await redis;
  await rc.quit();
  return;
}
