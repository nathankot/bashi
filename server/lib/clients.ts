import * as t from "io-ts";
import * as r from "redis";
import { Configuration, OpenAIApi } from "openai";

import "dotenv/load.ts";

const redisUrl = Deno.env.get("REDIS_URL");
if (redisUrl == null || redisUrl.length === 0) {
  throw new Error("REDIS_URL must be declared");
}

export async function withRedis<T>(
  f: (client: ReturnType<typeof r.createClient>) => Promise<T>
): Promise<T> {
  const client = r.createClient({
    url: Deno.env.get("REDIS_URL"),
  });

  let result: T;
  await client.connect();
  try {
    result = await f(client);
  } finally {
    await client.disconnect();
  }

  return result;
}

export const openai = new OpenAIApi(
  new Configuration({
    apiKey: Deno.env.get("OPENAI_KEY"),
  })
);

const whisperEndpoint = Deno.env.get("WHISPER_TRANSCRIBE_ENDPOINT")!;
if (whisperEndpoint == null || whisperEndpoint.length === 0) {
  throw new Error("WHISPER_TRANSCRIBE_ENDPOINT must be declared");
}

export { whisperEndpoint };

const GoogleSearchResponse = t.partial({
  items: t.array(
    t.type({
      title: t.string,
      link: t.string,
      snippet: t.string,
    })
  ),
});

export async function googleSearch(
  q: string,
  signal?: AbortSignal
): Promise<
  {
    link: string;
    title: string;
    snippet: string;
  }[]
> {
  const key = Deno.env.get("GOOGLE_SEARCH_API_KEY");
  const cx = Deno.env.get("GOOGLE_SEARCH_ENGINE_ID");
  if (key == null) {
    throw new Error(`GOOGLE_SEARCH_API_KEY is missing`);
  }
  if (cx == null) {
    throw new Error(`GOOGLE_SEARCH_ENGINE_ID is missing`);
  }
  const response = await fetch(
    "https://www.googleapis.com/customsearch/v1?" +
      new URLSearchParams({ key, cx, q, num: "5" }),
    { signal }
  );
  const json = await response.json();
  if (!GoogleSearchResponse.is(json)) {
    throw new Error("could not parse response from google search");
  }
  return (
    json.items?.map((r) => ({
      title: r.title ?? "<no title>",
      link: r.link ?? "<no link>",
      snippet: r.snippet ?? "<no snippet>",
    })) ?? []
  );
}
