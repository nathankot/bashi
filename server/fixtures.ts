import { Session } from "@/types.ts";
import { HandlerContext } from "$fresh/server.ts";

export const handlerCtx: HandlerContext<{}> = {
  params: {},
  render: (data) => new Response(JSON.stringify(data)),
  renderNotFound: () => new Response("", { status: 404 }),
  state: {},
  localAddr: {
    transport: "tcp",
    hostname: "127.0.0.1",
    port: 80,
  },
  remoteAddr: {
    transport: "tcp",
    hostname: "127.0.0.1",
    port: 80,
  },
};

export const session: Session = {
  commands: [],
  expiresAt: new Date(new Date().getTime() + 60000),
  sessionId: "a9fb6273-00ee-4e4c-9918-e87e1157ca31",
};
