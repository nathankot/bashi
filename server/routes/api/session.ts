import { Handlers } from "$fresh/server.ts";
import { renderError, renderJSON } from "@/util.ts";
import * as t from "io-ts";
import * as f from "fp-ts";

// import { renderError } from "@/util.ts";

const PostSessionRequest = t.type({
  commands: t.array(
    t.type({
      name: t.string,
      description: t.string,
      args: t.array(
        t.type({
          name: t.string,
          type: t.keyof({
            string: null,
            number: null,
            boolean: null,
          }),
        })
      ),
    })
  ),
});

interface PostSessionResponse {
  sessionId: string;
}

export const handler: Handlers<PostSessionResponse> = {
  async POST(req, ctx) {
    // TODO validate api key, for now just require it in the header.

    let json;
    try {
      json = await req.json();
    } catch {
      return renderError(400, "could not parse json");
    }
    const decodeResult = PostSessionRequest.decode(json);
    if (f.either.isLeft(decodeResult)) {
      return renderError(400, "malformed request");
    }

    // DONT TRUST THE USER.

    const decoded = decodeResult.right;
    console.log(decoded);

    const sessionId = crypto.randomUUID();
    return renderJSON<PostSessionResponse>({
      sessionId,
    });
  },
};
