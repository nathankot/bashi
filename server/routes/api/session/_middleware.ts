import * as f from "fp-ts";

import { MiddlewareHandlerContext } from "$fresh/server.ts";
import { Buffer } from "std/node/buffer.ts";

import msgpack from "@/msgpack.ts";
import { Session } from "@/types.ts";
import { withRedis } from "@/clients.ts";
import { renderError } from "@/util.ts";

export interface State {
  session?: Session;
}

export async function handler(
  req: Request,
  ctx: MiddlewareHandlerContext<State>
) {
  const authorization = req.headers.get("Authorization");
  if (authorization == null) {
    return renderError(401, "no Authorization header found");
  }

  if (!authorization.startsWith("Bearer ")) {
    return renderError(401, "malformed Authorization header");
  }

  const sessionId = authorization.substring("Bearer ".length);

  const bin: Buffer | null = await withRedis((client) =>
    client.get(
      client.commandOptions({
        returnBuffers: true,
      }),
      "s:" + sessionId
    )
  );

  if (bin == null) {
    return renderError(401, "session not found or expired");
  }

  const deserialized = msgpack.deserialize(bin.buffer);
  const decoded = Session.decode(deserialized);

  if (f.either.isLeft(decoded)) {
    return renderError(500, "internal error");
  }

  const session: Session = decoded.right;
  ctx.state.session = session;

  const resp = await ctx.next();
  return resp;
}
