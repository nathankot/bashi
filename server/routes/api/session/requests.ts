import * as f from "fp-ts";

import { Handlers } from "$fresh/server.ts";

import { renderError, renderJSON, handleError } from "@lib/util.ts";

import { State as ApiState } from "@routes/api/_middleware.ts";
import { State } from "./_middleware.ts";
import { Input, Output, run, validateInput } from "@lib/models.ts";
import interceptors from "@lib/output_interceptors.ts";

export type PostRequestsRequest = Input;
export type PostRequestsResponse = Output;

export const handler: Handlers<Output, State & ApiState> = {
  async POST(req, ctx) {
    let json;
    try {
      json = await req.json();
    } catch {
      return renderError(400, "could not parse json");
    }

    const inputDecodeResult = Input.decode(json);
    if (!f.either.isRight(inputDecodeResult)) {
      return renderError(400, "malformed request");
    }

    const input: Input = inputDecodeResult.right as any;
    const modelDeps = { openai: ctx.state.clients.openai };
    const modelName = input.model;

    try {
      switch (modelName) {
        case "assist-davinci-003":
        case "noop":
          if (!validateInput(modelName, input)) {
            throw new Error("input could not be verified");
          }
          let output = await run(
            modelName,
            modelDeps,
            ctx.state.session,
            input
          );

          for (const interceptor of interceptors) {
            output = await interceptor(ctx.state.session, output);
          }

          return renderJSON(output);
        default:
          const exhaustiveCheck: never = modelName;
          throw new Error(`model ${exhaustiveCheck} not found`);
      }
    } catch (e) {
      return handleError(e);
    }
  },
};
