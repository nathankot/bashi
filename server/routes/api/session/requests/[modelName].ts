import * as f from "fp-ts";

import { Handlers } from "$fresh/server.ts";

import HTTPError from "@lib/http_error.ts";
import { renderError, renderJSON, handleError } from "@lib/util.ts";
import { wrap } from "@lib/log.ts";
import { defaultPolicy } from "@lib/faultHandling.ts";
import {
  AllInput,
  AllOutput,
  ModelName,
  InputFor,
  ModelDeps,
  run,
} from "@lib/models.ts";

import { State as ApiState } from "@routes/api/_middleware.ts";
import { State } from "../_middleware.ts";

export type Request = FormData | Omit<AllInput, "model">;

export const handler: Handlers<AllOutput, State & ApiState> = {
  async POST(req, ctx) {
    let log = ctx.state.log;

    const modelNameDecodeResult = ModelName.decode(ctx.params["modelName"]);
    if (!f.either.isRight(modelNameDecodeResult)) {
      return handleError(log, new HTTPError(`model name invalid`, 400));
    }
    const model: ModelName = modelNameDecodeResult.right as any;
    log = wrap({ model }, ctx.state.log);

    const modelDeps: Omit<ModelDeps, "signal"> = {
      log,
      openai: ctx.state.clients.openai,
      whisperEndpoint: ctx.state.clients.whisperEndpoint,
      session: ctx.state.session,
      now: ctx.state.now,
      faultHandlingPolicy: defaultPolicy,
    };

    try {
      switch (model) {
        case "whisper":
          const arrayBuffer = await req.arrayBuffer();
          if (arrayBuffer == null) {
            return renderError(400, "no audio found in the request body");
          }
          const arrayBufferInput: InputFor<typeof model> = {
            model,
            arrayBuffer,
          };

          return renderJSON(await run(modelDeps, model, arrayBufferInput));

        case "assist-000":
        case "translate-davinci-003":
        case "code-000":
        case "noop":
          // JSON request
          let json;
          try {
            json = await req.json();
          } catch {
            return renderError(400, "could not parse json");
          }

          const inputDecodeResult = AllInput.decode({ model, ...json });
          if (!f.either.isRight(inputDecodeResult)) {
            return renderError(400, "malformed request");
          }

          const requestInput: InputFor<typeof model> =
            inputDecodeResult.right as any;

          if (
            "request" in requestInput &&
            requestInput.request.trim().length === 0
          ) {
            return renderError(400, "request must not be empty");
          }

          return renderJSON(await run(modelDeps, model, requestInput));

        default:
          const exhaustiveCheck: never = model;
          throw new Error(`model ${exhaustiveCheck} not found`);
      }
    } catch (e) {
      log("error", e);
      return handleError(log, e);
    }
  },
};
