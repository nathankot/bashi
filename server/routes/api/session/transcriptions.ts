import { Handlers } from "$fresh/server.ts";

import { State } from "@routes/api/_middleware.ts";
import { handleError, renderJSON, renderError } from "@lib/util.ts";

interface PostTranscriptionsResponse {
  text: string;
}

export const handler: Handlers<PostTranscriptionsResponse, State> = {
  async POST(req, ctx) {
    let audio = await req.arrayBuffer();
    if (audio == null) {
      return renderError(400, "no audio found in the request body");
    }
    let text: string;
    try {
      text = await ctx.state.clients.whisper.transcribe(ctx.state.log, audio);
    } catch (e) {
      ctx.state.log("error", e);
      return handleError(e);
    }
    return renderJSON<PostTranscriptionsResponse>({ text });
  },
};
