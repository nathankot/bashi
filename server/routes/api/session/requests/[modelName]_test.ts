import { assertSnapshot } from "std/testing/snapshot.ts";
import { stub } from "std/testing/mock.ts";

import { Session } from "@lib/session.ts";
import * as fixtures from "@lib/fixtures.ts";
import * as clients from "@lib/clients.ts";
import { handler } from "./[modelName].ts";

const nowFixture = fixtures.now;

for (const test of [
  {
    description: "success",
    model: "assist-000",
    request: `{ "request": "whats the time in new york? email translate" }`,
    session: {
      ...fixtures.session,
      configuration: {
        ...fixtures.session.configuration,
        disabledBuiltinFunctions: ["translate"],
      },
    } as Session,
    openAiFn: async () =>
      ({
        data: {
          choices: [
            {
              text: `
answer("mock response")
   unknownCall(123)
unparseable(
translate("English", "should not work because disabled by conf")
time("America/New_York")
sendEmail("not enough args")
sendEmail("blah@blah.com", "some subject", "some body")
flushToSpeech()
reminder("in 5 days", "some reminder name")`,
            },
          ],
        },
      } as any),
  },
  {
    description: "openai errors",
    model: "assist-000",
    request: `{ "request": "whats the time in new york?" }`,
    session: fixtures.session,
    openAiFn: async () => {
      throw new Error("mock error");
    },
  },
  {
    description: "empty request",
    model: "assist-000",
    request: `{ "request": "" }`,
    session: fixtures.session,
  },
  {
    description: "noop model works",
    model: "noop",
    request: `{ "request": "noop" }`,
    session: fixtures.session,
    openAiFn: null,
  },
  // missing request context
  // adding request context
  // too many tokens
  // rate limited (or maybe handled by middleware?)
]) {
  Deno.test(
    `POST /api/session/requests/${test.model}: ` + test.description,
    async (t) => {
      const openAiStub =
        test.openAiFn == null
          ? null
          : stub(clients.openai, "createCompletion", test.openAiFn);

      try {
        const enc = new TextEncoder();
        const req = new Request(
          `http://localhost/api/session/requests/${test.model}`,
          {
            method: "POST",
            body: enc.encode(test.request),
          }
        );
        req.headers.set("Content-Type", "application/json");
        const response = await handler.POST!(req, {
          ...fixtures.handlerCtx,
          params: {
            modelName: test.model,
          },
          state: {
            now: () => nowFixture,
            clients,
            session: test.session,
            log: () => {},
          },
        });
        const responseJson = await response.json();
        await assertSnapshot(t, responseJson);
      } finally {
        if (openAiStub != null) {
          openAiStub.restore();
        }
      }
    }
  );
}
