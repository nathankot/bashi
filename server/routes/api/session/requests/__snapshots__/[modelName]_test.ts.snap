export const snapshot = {};

snapshot[`POST /api/session/requests/assist-000: success 1`] = `
{
  functionCalls: [
    {
      args: [
        "mock response",
      ],
      argsParsed: [
        {},
      ],
      line: 'answer("mock response")',
      name: "answer",
      type: "parsed",
    },
    {
      args: [
        123,
      ],
      invalidReason: "unknown_function",
      line: "   unknownCall(123)",
      name: "unknownCall",
      type: "invalid",
    },
    {
      error: "<END-OF-FILE>: Unable to consume token: <END-OF-FILE>",
      line: "unparseable(",
      type: "parse_error",
    },
    {
      args: [
        "English",
        "should not work because disabled by conf",
      ],
      invalidReason: "unknown_function",
      line: 'translate("English", "should not work because disabled by conf")',
      name: "translate",
      type: "invalid",
    },
    {
      args: [
        "America/New_York",
      ],
      argsParsed: [
        {},
      ],
      line: 'time("America/New_York")',
      name: "time",
      returnValue: {
        type: "string",
        value: "12/19/2022, 3:41:10 AM",
      },
      type: "executed",
    },
    {
      args: [
        "not enough args",
      ],
      invalidReason: "invalid_arguments",
      line: 'sendEmail("not enough args")',
      name: "sendEmail",
      type: "invalid",
    },
    {
      args: [
        "blah@blah.com",
        "some subject",
        "some body",
      ],
      argsParsed: [
        {},
        {},
        {},
      ],
      line: 'sendEmail("blah@blah.com", "some subject", "some body")',
      name: "sendEmail",
      type: "parsed",
    },
    {
      args: [
      ],
      argsParsed: [
      ],
      line: "flushToSpeech()",
      name: "flushToSpeech",
      type: "parsed",
    },
    {
      args: [
        "in 5 days",
        "some reminder name",
      ],
      invalidReason: "unknown_function",
      line: 'reminder("in 5 days", "some reminder name")',
      name: "reminder",
      type: "invalid",
    },
  ],
  model: "assist-000",
  request: "whats the time in new york? email translate",
}
`;

snapshot[`POST /api/session/requests/assist-000: openai errors 1`] = `
{
  error: "internal server error",
}
`;

snapshot[`POST /api/session/requests/assist-000: empty request 1`] = `
{
  error: "request must not be empty",
}
`;

snapshot[`POST /api/session/requests/noop: noop model works 1`] = `
{
  model: "noop",
  result: "noop",
}
`;
